### PR TITLE
Refactor and enhance Association and Attributes management

### PR DESCRIPTION
**Pull request by Dragonfly911117**
### Description

This pull request includes significant refactorings and enhancements related to the Association and AssAttribute components:

- Converted `AssAttribute.ratio` to public and renamed it to `Ratio` for better accessibility.
- Enhanced `Association.UpdateDrawData` with improved validation for nil associations, parents, and attributes.
- Introduced stricter input validation for `AssociationType` and `Ratio` boundaries.
- Simplified and standardized various struct definitions and methods for clarity and consistency.
- Reorganized and categorized methods in the `Association` component for better readability.
- Added new methods to manage attributes, layers, and association types in the Association struct.
- Included utility methods for draw data updates and retrieval.
- Refactored and removed redundant or duplicate methods.
- Improved error handling, removed unnecessary return values, and extracted coordinate-related fields for alignment with existing conventions.

Additionally, unit test coverage has been improved to cover edge cases and handle validation scenarios comprehensively.

### Checklist

- [x] Code compiles correctly
- [x] Tests for the changes have been added (for bug fixes/features)
- [x] All tests passing
- [ ] Coding standards followed

### CODE REVIEW

File: app/backend/component/association.go [hunk 1]
```diff
@@ -1,82 +1,146 @@
...
-		parents: [2]*Gadget{parents[0], parents[1]},
-	}, nil
+	this.attributes = append(this.attributes, attribute)
+	return nil
 }
 
-func (a *Association) GetParentStart() (*Gadget, duerror.DUError) {
-	return a.parents[0], nil
+func (this *Association) Cover(p utils.Point) (bool, duerror.DUError) {
+	return false, nil
```
Comment (UGisBusy) [inline app/backend/component/association.go:105]: 幫我加個 TODO

File: app/backend/component/gadget.go [hunk 2]
```diff
@@ -126,3 +127,9 @@ func NewGadget(gadgetType GadgetType, point utils.Point) (*Gadget, duerror.DUErr
 	}
 	return &g, nil
 }
+
+func (g *Gadget) getBounds() (utils.Point, utils.Point, duerror.DUError) {
+	// TODO: calculate the Bottom-Right point (maybe store it?)
+	size := 5
+	return g.point, utils.AddPoints(g.point, utils.Point{X: size, Y: size}), nil
+}
```
Comment (UGisBusy) [inline app/backend/component/gadget.go:135]: remove these

File: app/backend/components/components.go [hunk 3]
```diff
@@ -111,3 +111,43 @@ func (cs *Components) RegisterUpdateParentDraw(update func() duerror.DUError) du
 	cs.updateParentDraw = update
 	return nil
 }
+
+func (cs *Components) AddAssociation(assType component.AssociationType, point [2]utils.Point) duerror.DUError {
+	if point[0].X < 0 || point[0].Y < 0 || point[1].X < 0 || point[1].Y < 0 {
+		return duerror.NewInvalidArgumentError("points coordinates must be non-negative")
+	}
+	var parents [2]*component.Gadget
+	candidate, err := cs.componentsContainer.Search(point[0])
```
Comment (UGisBusy) [inline app/backend/components/components.go:120]:
```suggestion
	candidate, err := cs.componentsContainer.Search(point[0])
	// TODO: make container search only gadget
```

**Review by UGisBusy — APPROVED** [review]

**Review by UGisBusy — APPROVED** [review]

**Review by Vghxv — APPROVED** [review]
yolo

**Comment by Dragonfly911117** [comment]
@UGisBusy @Vghxv @garry911128 Review the title and desc again
